Allow the output address owner to change the output address via MsgStake

msmania opened the pull request:
This patch allows the owner of the output address of a staked node to change the output address to another via the existing stake transaction.  The same idea was proposed as #1499.

### Rationale:

This feature gives users more flexibility.  One of the possible use cases is that a node runner can convert custodial stake nodes into non-custodial stake nodes without unstaking.  Or a current non-custodial stake user may want to change the output address some time after they stake nodes.

### Technical details:

Below are the key functions in the transaction processing flow.  This patch makes a new `MsgStake` transaction to change the output address pass through all Validate* functions and lets the final function `EditStakeValidator` accept the change in the world state if the new feature key “OEDIT” is enabled.

1. BaseApp.runTx (baseapp.go)
    1. auth.ValidateTransaction (ante.go)
    2. BaseApp.runMsg (baseapp.go)
        1. nodes.handleStake (handler.go)
            1. keeper.ValidateValidatorStaking (valStateChanges.go)
                1. keeper.ValidateValidatorMsgSigner (for the new validator)
                2. keeper.ValidateValidatorMsgSigner (for the current validator)
                3. keeper.ValidateEditStake
            2. keeper.StakeValidator
                1. keeper.EditStakeValidator

The challenge was in `ValidateTransaction`, which expects a transaction message to be signed by a key which appears in the transaction itself.  For example, `MsgStake` must be signed by `MsgStake.PublicKey` or `MsgStake.Output`.  Since we intend to allow only the current output address owner to change the output address, however, a new `MsgStake` transaction must be signed by the current output address, and it doesn’t appear in the transaction because a new output address is there.  This patch needs to change that signer restriction rule.

Each message type implements `GetSigners` that returns a slice of `Address`.  When a transaction arrives at `ValidateTransaction`, it calls `GetSigners` and checks if a transaction is signed by any of the addresses returned from `GetSigners`.  If not, the transaction is rejected with `CodeUnauthorized`.  The challenge here is that `ValidateTransaction` is in the auth package that doesn’t know about `MsgStake`.  One possible solution is to make `MsgStake.GetSigners` return three addresses, `MsgStake.PublicKey`, `MsgStake.Output`, and the current output address.  However, implementing it is not straightforward because `MsgStake` is in the types package which doesn’t have direct access to the keeper package that is required to get the current output address.

The proposed solution in this patch is to introduce a new method `GetMsgStakeOutputSigner` in the keeper package and `ValidateTransaction` calls it to append the current output address to the slice returned from `GetSigners`.

The changes in valStateChanges.go are straightforward.  `ValidateValidatorMsgSigner` simply checks if a signer is either the node’s operator or output address.  We need to skip the call to `ValidateValidatorStaking` with the new validator because the signer does not appear in the new validator.  The final validate function `ValidateEditStake` ensures that `MsgStake` to change the output address must be signed by the current output address.  Otherwise it returns a new error `ErrDisallowedOutputAddressChange`.

After all validate functions, `EditStakeValidator` accepts a new output address.

### Testing:

1. Prepare a local network with NCUST enabled.
2. Set the following environment variables accordingly.
    ```
    export POKT=<path to pocket executable>
    export CHAINS=<chain IDs to host>
    export URL=<service url>
    export STAKE=<upokt to stake>
    export NETWORK=<network name>
    export NODE=<node address>
    export PUBKEY=<node public key>
    export PRIVKEY=<node private key>
    export OUTPUT1=<output address>
    export OUTPUT1_PRIVKEY=<OUTPUT1’s private key>
    export OUTPUT2=<another output address>
    export OUTPUT2_PRIVKEY=<OUTPUT2’s private key>
    export DAO=<dao address>
    ```
3. Make sure all addresses are stored in the keybase and have enough tokens.
4. Submit a non-custodial stake tx to stake a new node.
    ```
    $POKT accounts delete $OUTPUT1
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT1 $STAKE $CHAINS $URL $NETWORK 10000 false
    ```
5. Submit three stake transactions to change the output address signed by the new output address, the operator address, and the current output address respectively.  Verify all transactions are rejected with CodeUnauthorizedSigner (pos:125), CodeUnequalOutputAddr (pos:124), and CodeUnauthorized (sdk:4).
    ```
    $POKT accounts import-raw $OUTPUT1_PRIVKEY
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $OUTPUT2
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $NODE
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    ```
6. Enable OEDIT.  Choose the upgrade height accordingly.
    ```
    $POKT gov enable $DAO <upgrade height> OEDIT $NETWORK 10000
    ```
7. Submit three stake transactions to change the output address signed by the new output address, the operator address, and the current output address respectively.  Verify the first two transactions are rejected with CodeUnauthorizedSigner (pos:125) and CodeDisallowedOutputAddressEdit (pos:127) respectively, and the last transaction is accepted.
    ```
    $POKT accounts import-raw $PRIVKEY
    $POKT accounts import-raw $OUTPUT2_PRIVKEY
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $OUTPUT2
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    $POKT accounts delete $NODE
    $POKT nodes stake non-custodial $PUBKEY $OUTPUT2 $STAKE $CHAINS $URL $NETWORK 10000 false
    ```
8. Verify the node’s output address is changed to $OUTPUT2.

Olshansk commented on codec/codec.go:
```diff
@@ -266,6 +267,11 @@ func (cdc *Codec) IsAfterNonCustodialUpgrade(height int64) bool {
 
+func (cdc *Codec) IsAfterOutputAddressChangeUpgrade(height int64) bool {
+	return UpgradeFeatureMap[OutputAddressEditKey] != 0 &&
+		height >= UpgradeFeatureMap[OutputAddressEditKey]
```
Add a `|| TestMode <= -3` similar to the other recent changes

Olshansk commented on codec/codec.go:
```diff
@@ -266,6 +267,11 @@ func (cdc *Codec) IsAfterNonCustodialUpgrade(height int64) bool {
 	return (UpgradeFeatureMap[NonCustodialUpdateKey] != 0 && height >= UpgradeFeatureMap[NonCustodialUpdateKey]) || TestMode <= -3
 }
 
+func (cdc *Codec) IsAfterOutputAddressChangeUpgrade(height int64) bool {
```
Optional: s/IsAfterOutputAddressChangeUpgrade/IsAfterOutputAddressEditorUpgrade

wdyt?

Olshansk commented on app/app.go:
```diff
@@ -87,6 +87,7 @@ func NewPocketCoreApp(genState GenesisState, keybase keys.Keybase, tmClient clie
 	// give pocket keeper to nodes module for easy cache clearing
 	app.nodesKeeper.PocketKeeper = app.pocketKeeper
 	app.appsKeeper.PocketKeeper = app.pocketKeeper
+	app.accountKeeper.POSKeeper = app.nodesKeeper
```
For my personal understanding, can you explain to me why this was necessary?

msmania replied:
This is for cross-package keeper access.  What we need to achieve is to retrieve the current output address in `ValidateTransaction`, which needs `nodes/Keeper`.  `ValidateTransaction` is in the `auth` package, however, the keeper `ValidateTransaction` takes as a parameter is `auth/Keeper`.  This property `POSKeeper` gives us access to `nodes/Keeper`.

Olshansk commented on x/auth/ante.go:
```diff
@@ -61,8 +62,16 @@ func ValidateTransaction(ctx sdk.Ctx, k Keeper, stdTx types.StdTx, params Params
 		return nil, types.ErrDuplicateTx(ModuleName, hex.EncodeToString(txHash))
 	}
+
+	signersAllowed := stdTx.GetSigners()
```
s/signersAllowed/validSigners

Olshansk commented on x/auth/ante.go:
```diff
@@ -61,8 +62,16 @@ func ValidateTransaction(ctx sdk.Ctx, k Keeper, stdTx types.StdTx, params Params
 	if res != nil {
```
The intetion of this PR is to allow changes in the `MsgStake` command, but this validates ANY transaction and we are hacking the `MsgStake` piece via `GetExternalSigners`.

I personally would've looked at `x/nodes.handler.go` and made the change in:

```
func handleStake(ctx sdk.Ctx, msg types.MsgStake, k kee
```

This would allow removing the new interface altogether. 

Let me know if I'm missing something or what the thought process was.

msmania replied:
Before reaching `handleStake`, all transactions must pass the checks in `auth.ValidateTransaction`.  So we need to change that function.  I'll update the description with more details.

Olshansk commented on x/nodes/types/errors.go:
```diff
@@ -11,32 +10,33 @@ import (
+	CodeUnequalOutputAddr             CodeType          = 124
+	CodeUnauthorizedSigner            CodeType          = 125
+	CodeNilSigner                     CodeType          = 126
+	CodeDisallowedOutputAddressChange CodeType          = 127
```
I think this is more descriptive: `s/CodeDisallowedOutputAddressChange/CodeDisallowedOutputAddressEdit`

Olshansk commented on x/nodes/keeper/valStateChanges.go:
```diff
@@ -109,48 +109,58 @@ func (k Keeper) UpdateTendermintValidators(ctx sdk.Ctx) (updates []abci.Validato
-		return err
+func (k Keeper) ValidateValidatorStaking(ctx sdk.Ctx,
+	validatorNew types.Validator, amount sdk.BigInt,
+	signerAddress sdk.Address) sdk.Error {
```
```suggestion
func (k Keeper) ValidateValidatorStaking(
	ctx sdk.Ctx,
	validatorNew types.Validator,
	amount sdk.BigInt,
	signerAddress sdk.Address
) sdk.Error {
```

Olshansk commented on x/nodes/keeper/common_test.go:
```diff
@@ -135,6 +133,17 @@ func sendFromModuleToAccount(t *testing.T, ctx sdk.Ctx, k *Keeper, module string
 }
 
+func fundAccount(ctx sdk.Context, k Keeper,
+	address sdk.Address, coin sdk.Coin) sdk.Error {
```
NIT: Either keep everything on one line or multiline the whole thing
```suggestion
func fundAccount(
	ctx sdk.Context,
	k Keeper,
	address sdk.Address,
	coin sdk.Coin
) sdk.Error {
```

Olshansk commented on x/nodes/keeper/common_test.go:
```diff
@@ -135,6 +133,17 @@ func sendFromModuleToAccount(t *testing.T, ctx sdk.Ctx, k *Keeper, module string
+	err := k.AccountKeeper.MintCoins(ctx, types.StakedPoolName, coins)
+	if err != nil {
+		return err
+	}
```
```suggestion
	if err := k.AccountKeeper.MintCoins(ctx, types.StakedPoolName, coins); err != nil
		return err
	}
```

Olshansk commented on x/nodes/keeper/common_test.go:
```diff
@@ -135,6 +133,17 @@ func sendFromModuleToAccount(t *testing.T, ctx sdk.Ctx, k *Keeper, module string
 	}
 }
 
+func fundAccount(ctx sdk.Context, k Keeper,
```
```suggestion
// fundAccount mints new tokens in the stake pool and sends them
// to the address provided, for testing purposes.
func fundAccount(ctx sdk.Context, k Keeper,
```

Olshansk commented on x/auth/types/expectedKeepers.go:
```diff
@@ -0,0 +1,9 @@
+)
+
+type PosKeeper interface {
+	GetExternalSigners(sdk.Ctx, sdk.Msg) []sdk.Address
```
I left a comment above that might let us eliminate the need for this. 

But, if we do keep it: `GetMsgStakeOutputSigner(sdk.Ctx, sdk.Msg) *sdk.Address`

1. Make it clear and explicit
2. There will only ever be one in V0

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
 	}
 }
 
+func runStake(ctx sdk.Context,
```
Ditto msg header

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+		msg.Output)
+	err := k.ValidateValidatorStaking(
+		ctx, validator, msg.Value, sdk.Address(signer.Address()))
+	if err != nil {
```
```suggestion
	if err := k.ValidateValidatorStaking(ctx, validator, msg.Value, sdk.Address(signer.Address())); err != nil {
```

Olshansk commented on x/nodes/keeper/valStateChanges.go:
```diff
@@ -109,48 +109,58 @@ func (k Keeper) UpdateTendermintValidators(ctx sdk.Ctx) (updates []abci.Validato
+	// check to see if the public key has already been register for that validator
+	validatorCur, found := k.GetValidator(ctx, validatorNew.Address)
+	if found &&
+		validatorNew.Address.Equals(validatorCur.Address) &&
```
Can you move the formulation of this boolean into a local var.

Then the if will be simple to read and you can add `NOOP` comment inside.

Olshansk commented on x/nodes/keeper/valStateChanges.go:
```diff
@@ -109,48 +109,58 @@ func (k Keeper) UpdateTendermintValidators(ctx sdk.Ctx) (updates []abci.Validato
+		k.Cdc.IsAfterOutputAddressChangeUpgrade(ctx.BlockHeight()) {
+		// In this case, for edit stake, we skip verifying an operation
+		// against the new validator to accept a signer that appears
+		// only in the current validator.
```
Doesn't the `outputAddress` need to be the `signerAddress`?

msmania replied:
Yes, I delegated that check to `ValidateEditStake`, but having it here too does no harm.  Will add it.

Olshansk commented on x/nodes/keeper/valStateChanges.go:
```diff
@@ -228,10 +238,18 @@ func (k Keeper) ValidateEditStake(ctx sdk.Ctx, currentValidator, newValidtor typ
+			// ensure only the current output address owner can change the output address
+			if currentValidator.OutputAddress != nil &&
+				!signer.Equals(currentValidator.OutputAddress) &&
+				!newValidtor.OutputAddress.Equals(currentValidator.OutputAddress) {
```
Wouldn't `!newValidtor.OutputAddress.Equals(currentValidator.OutputAddress)` prevent the signer from changing the output address altogether?

msmania replied:
If a signer is not the current output address, yes, we prevent the output address change.

Olshansk commented on x/nodes/keeper/valStateChanges.go:
```diff
@@ -228,10 +238,18 @@ func (k Keeper) ValidateEditStake(ctx sdk.Ctx, currentValidator, newValidtor typ
+			}
+		} else {
+			//check the "new" validator's signature validity
+			//will recheck if validator exists
```
Why not keep this as `// ensure output address doesn't change` like before? It's pretty clear.
```suggestion
			// check the "new" validator's signature validity
			// will re-check if validator exists
```

Olshansk commented on x/nodes/keeper/valStateChanges.go:
```diff
@@ -109,48 +109,58 @@ func (k Keeper) UpdateTendermintValidators(ctx sdk.Ctx) (updates []abci.Validato
-		if ctx.IsAfterUpgradeHeight() && val.IsStaked() {
-			return k.ValidateEditStake(ctx, val, validator, amount, signerAddress)
+		if ctx.IsAfterUpgradeHeight() && validatorCur.IsStaked() {
+			return k.ValidateEditStake(ctx, validatorCur, validatorNew, amount, signerAddress)
```
This feature is all about `editing` stake, so why not move the core business logic (with the feature checks) into `ValidateEditStake`? 

I think it'll be easier to read that way.

msmania replied:
Before `ValidateEditStake`, `ValidateValidatorStaking` calls `ValidateValidatorMsgSigner` twice, one with the new validator and one with the current validator.  We need to accept a stake transaction signed by the current output address, but it doesn't pass the first `ValidateValidatorMsgSigner` against the new validator because the new validator doesn't have the signer, the current output address.  That's why we need a logic to skip the first `ValidateValidatorMsgSigner`.  We can't move that part to `ValidateEditStake`.

Well, I'd like to remove `ValidateValidatorMsgSigner` and do all checks inside `ValidateEditStake`, but that kind of refactoring change shouldn't be included in this patch.

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	return k.StakeValidator(ctx, validator, msg.Value, signer)
+}
+
+func Test_OutputAddressChange(t *testing.T) {
```
TestValidatorStateChange_OutputAddressEdit

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	codec.UpgradeHeight = -1
+
+	stakeAmount := sdk.NewCoin(k.StakeDenom(ctx), sdk.NewInt(k.MinimumStake(ctx)))
+	outputAccount := getRandomPubKey()
```
s/outputAccount/outputPubKey

Olshansk replied:
Ditto below

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	stakeAmount := sdk.NewCoin(k.StakeDenom(ctx), sdk.NewInt(k.MinimumStake(ctx)))
+	outputAccount := getRandomPubKey()
+	outputAddress := sdk.Address(outputAccount.Address())
+	msgTemplate := types.MsgStake{
```
s/msgTemplate/msgStakeTemplate

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	assert.Nil(t, fundAccount(ctx, k, outputAddress, stakeAmount))
+
+	// Stake two nodes before NCUST
+	msgTemplate.PublicKey = operatorAcc1
```
I think you should copy `msgStake = proto.Clone(msgStakeTemplate)` and then update it.

It doesn't matter from a test passing perspective, but just makes more sense and less prone to error.

Olshansk replied:
Ditto below

msmania replied:
Yeah, if it's named `msgTemplate`, it should behave as a template.  I found a better way to call `runStake`.

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	assert.Nil(t, runStake(ctx, k, msgTemplate, operatorAcc1))
+	validatorCur, found = k.GetValidator(ctx, operatorAddr1)
+	assert.True(t, found)
+	assert.True(t, validatorCur.OutputAddress.Equals(outputAddress))
```
Use `assert.Equal` directly

msmania replied:
Oh I didn't know `assert.Equal` is that smart.  Good to know.

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	// Attempt to change the output address with operator's signature --> Fail
+	msgTemplate.PublicKey = operatorAcc1
+	msgTemplate.Output = operatorAddr1
+	err := runStake(ctx, k, msgTemplate, operatorAcc1)
```
assert that it's NotNil first. Ditto below a few times

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	assert.True(t, found)
+	assert.True(t, validatorCur.OutputAddress.Equals(operatorAddr1))
+
+	// Attempt to change the output address from nil
```
This change is reasonable but wasn't clear from the github description. Please update it appropriately.

Olshansk commented on x/nodes/keeper/valStateChanges_test.go:
```diff
@@ -350,6 +352,129 @@ func TestValidatorStateChange_EditAndValidateStakeValidatorAfterNonCustodialUpgr
+	validatorCur, found = k.GetValidator(ctx, operatorAddr3)
+	assert.True(t, found)
+	assert.True(t, validatorCur.OutputAddress.Equals(outputAddress))
+}
```
Can you please update the github description with:
1. A list of all the changes
2. Example `pocket` commands you ran to verify this on a locally running node
3. Just two sentences on a reasoning / rationale that we'll include in the release notes

msmania committed:
Allow output address change
This patch allows the owner of the output address of a staked node to change
the output address to another via MsgStake after the new feature key "OEDIT"
is enabled.

In addition to the server side change, this patch updates the behavior of
the `pocket nodes stake non-custodial` command to get an MsgStake to be signed
by the current output address if neither the operator nor the new output
address is not available in the key base.

Olshansk commented:
@msmania 

Here's a mermaid diagram ChatGPT gave me using your github description.

<img width="548" alt="Screenshot 2023-04-04 at 7 44 37 PM" src="https://user-images.githubusercontent.com/1892194/229967871-74c492fc-1b2a-417e-bc3a-49a5d1ad1492.png">

<img width="536" alt="Screenshot 2023-04-04 at 7 44 02 PM" src="https://user-images.githubusercontent.com/1892194/229967800-73aabf61-ac9a-4a0c-a65d-6820d1abe99b.png">

```mermaid
flowchart TD
    A(BaseApp.runTx) --> B(auth.ValidateTransaction)
    B --> C(BaseApp.runMsg)
    C --> D(nodes.handleStake)
    D --> E(keeper.ValidateValidatorStaking)
    E --> F(keeper.ValidateValidatorMsgSigner)
    F --> G(keeper.ValidateValidatorMsgSigner)
    G --> H(keeper.ValidateEditStake)
    H --> I(keeper.StakeValidator)
    I --> J(keeper.EditStakeValidator)

    subgraph Before
        B -->|Validate using signed keys|L(MsgStake.GetSigners)
        H -->|Ensure signed by current output address| M(Error)
    end

    subgraph After
        B --> K(GetMsgStakeOutputSigner)
        K -->|Append current output address|L(MsgStake.GetSigners)
        H -->|Ensure signed by current output address| M(ValidateEditStake)
    end
```

Olshansk commented on app/cmd/cli/txUtil.go:
```diff
@@ -123,11 +122,41 @@ func LegacyStakeNode(chains []string, serviceURL, fromAddr, passphrase, chainID
+	if err != nil {
+		return nil, err
+	}
+	var node nodeTypes.Validator
```
NIT: This should go above the unmarshal where you use it

Olshansk commented on app/cmd/cli/txUtil.go:
```diff
@@ -123,11 +122,41 @@ func LegacyStakeNode(chains []string, serviceURL, fromAddr, passphrase, chainID
 	}, nil
 }
 
+func getCurrentOutputAddress(operatorAddr sdk.Address) (sdk.Address, error) {
```
```suggestion
// getCurrentOutputAddress returns the current output address for the operator provided.
// nil is returned if there is no output address.
func getCurrentOutputAddress(operatorAddr sdk.Address) (sdk.Address, error) {
```

Olshansk commented on app/cmd/cli/txUtil.go:
```diff
@@ -123,11 +122,41 @@ func LegacyStakeNode(chains []string, serviceURL, fromAddr, passphrase, chainID
+		return nil, err
+	}
+
+	return node.OutputAddress, err
```
```suggestion
	return node.OutputAddress, nil
```

msmania replied:
oops, copy-n-paste mistake

Olshansk commented on app/cmd/cli/txUtil.go:
```diff
@@ -123,11 +122,41 @@ func LegacyStakeNode(chains []string, serviceURL, fromAddr, passphrase, chainID
+func getFirstAddressAvailableInKeybase(
+	kb keys.Keybase,
+	addresses []sdk.Address,
+) (sdk.Address, bool) {
```
Why not return an `*sdk.Address` so you can check for a nil or non-nil?

msmania replied:
Ah right, that's why you suggested a pointer.  I got it now.

Olshansk commented on app/cmd/cli/txUtil.go:
```diff
@@ -123,11 +122,41 @@ func LegacyStakeNode(chains []string, serviceURL, fromAddr, passphrase, chainID
+	return node.OutputAddress, err
+}
+
+func getFirstAddressAvailableInKeybase(
```
NIT: Can you add a comment on the use case for this? I kind of feel like you're use-case is more like `anyExistsInKeybase`

msmania replied:
I'd prefer to keep "First" in the name because the searching order is important.  Will add a leading comment.

Olshansk commented on app/cmd/cli/stake.go:
```diff
@@ -93,11 +94,19 @@ var nonCustodialstakeCmd = &cobra.Command{
+If the node is currently staked at X and you submit an update with new stake Y,
+only Y-X will be subtracted from an account.
+If no changes are desired for the parameter, just enter the current param value
+just as before.`,
```
```suggestion

If the node is currently staked at X and you submit an update with new stake Y,
only Y-X will be subtracted from an account.

If no changes are desired for the parameter, just enter the current param values as before.`,
```

Olshansk commented on x/nodes/keeper/valStateChanges.go:
```diff
@@ -109,48 +108,68 @@ func (k Keeper) UpdateTendermintValidators(ctx sdk.Ctx) (updates []abci.Validato
+	// new validator's state.  So we skip ValidateValidatorMsgSigner if 1) the
+	// current height is after the upgrade, 2) this is an edit-stake changing the
+	// output address, and 3) the signer is the current output address.
+	skipMsgSignerValidation := found &&
```
This is ultra clear and safe. Ty!!

Olshansk reviewed (changes requested):
A couple of minor NITS/comments, but otherwise this PR LGTM from a code perspective. Amazing test and amazingly clear code!

We need to submit a proposal to the DAO to approve this (cc @jessicadaugherty), but I don't expect any opposition from the community.

msmania committed:
Address nit comments

Olshansk reviewed (dismissed):
LGTM!! :shipit:

msmania commented:
> @msmania
> 
> Here's a mermaid diagram ChatGPT gave me using your github description.

Interesting.  Fortunately or unfortunately, AI made a mistake.  That flow is not correct.  I think it's hard to overlap before and after in the same chart.  Here are two charts before and after.  This matches the test scenario in the first description.

### Sending MsgStake without this patch

```mermaid
flowchart TD
  A(runTx) --> B(ValidateTransaction)
  B -- If signed by the current output address --> K([sdk:4])
  A --> C(runMsg)
  C --> D(handleStake)
  D --> E(ValidateValidatorStaking)
  D -.-> F(StakeValidator)
  E --> G(ValidateValidatorMsgSigner\nwith the current state)
  G -- If signed by the new output address --> L([pos:125])
  E --> H(ValidateValidatorMsgSigner\nwith the new state)
  E --> I(ValidateEditStake)
  I -- If signed by the operator --> M([pos:124])
  F -.-> J(EditStakeValidator)
```

### Sending MsgStake with this patch

```mermaid
flowchart TD
  A(runTx) --> B(ValidateTransaction)
  B --> N(GetMsgStakeOutputSigner)
  N -- If signed by the current output address --> K([ok])
  A --> C(runMsg)
  C --> D(handleStake)
  D --> E(ValidateValidatorStaking)
  D --> F(StakeValidator)
  E --> O{Valid output address edit?}
  O -. No .-> G(ValidateValidatorMsgSigner\nwith the current state)
  G -.-> L([pos:125])
  E --> H(ValidateValidatorMsgSigner\nwith the new state)
  E --> I(ValidateEditStake)
  I -- If signed by the operator --> M([pos:127])
  F --> J(EditStakeValidator)
  J --> P([Happy ending!])
```

msmania committed:
Merge branch 'staging' into feature-output-address-change

Olshansk commented:
@msmania Wanted to encourage you to submit another PR updating the documentation so these awesome diagrams don't go to waste!

msmania commented:
> @msmania Wanted to encourage you to submit another PR updating the documentation so these awesome diagrams don't go to waste!

Created #1544
